lib/internal/modules/esm/worker.js

```diff
@@ -0,0 +1,208 @@
+  if (isTypedArray(source)) return [TypedArrayPrototypeGetBuffer(source)];
+  if (isDataView(source)) return [DataViewPrototypeGetBuffer(source)];
+}
+
+function wrapMessage(status, body) {
+  if (status === 'success' || body === null || typeof body !== 'object') {
+    return { status, body };
+  }
+  let serialized;
+  let serializationFailed;
+  try {
+    const { serializeError } = require('internal/error_serdes');
+    serialized = serializeError(body);
+  } catch {
+    serializationFailed = true;
+  }
+  return {
+    status,
+    body: {
+      serialized,
+      serializationFailed,
+    },
+  };
+}
+
+async function customizedModuleWorker(lock, syncCommPort, errorHandler) {
+  let hooks, preloadScripts, initializationError;
+  let hasInitializationError = false;
+
+  try {
+    initializeESM();
+    const initResult = await initializeHooks();
+    hooks = initResult.hooks;
+    preloadScripts = initResult.preloadScripts;
+  } catch (exception) {
+    // If there was an error while parsing and executing a user loader, for example if because a
+    // loader contained a syntax error, then we need to send the error to the main thread so it can
+    // be thrown and printed.
+    hasInitializationError = true;
+    initializationError = exception;
+  }
+
+  syncCommPort.on('message', handleMessage);
+
+  if (hasInitializationError) {
+    syncCommPort.postMessage(wrapMessage('error', initializationError));
+  } else {
+    syncCommPort.postMessage(wrapMessage('success', { preloadScripts }), preloadScripts.map(({ port }) => port));
+  }
+
+  // We're ready, so unlock the main thread.
+  AtomicsAdd(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION, 1);
+  AtomicsNotify(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION);
+
+  let immediate;
+  function checkForMessages() {
+    immediate = setImmediate(checkForMessages).unref();
+    // We need to let the event loop tick a few times to give the main thread a chance to send
+    // follow-up messages.
+    const response = receiveMessageOnPort(syncCommPort);
+
+    if (response !== undefined) {
+      PromisePrototypeThen(handleMessage(response.message), undefined, errorHandler);
+    }
+  }
+
+  const unsettledResponsePorts = new SafeSet();
+
+  process.on('beforeExit', () => {
+    for (const port of unsettledResponsePorts) {
+      port.postMessage(wrapMessage('never-settle'));
+    }
+    unsettledResponsePorts.clear();
+
+    AtomicsAdd(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION, 1);
+    AtomicsNotify(lock, WORKER_TO_MAIN_THREAD_NOTIFICATION);
+
+    // Let's attach back the event handler.
+    syncCommPort.on('message', handleMessage);
+    // Let's also check synchronously for a message, in case it's already there.
+    clearImmediate(immediate);
+    checkForMessages();
+    // We don't need the sync check after this tick, as we already have added the event handler.
+    clearImmediate(immediate);
+    // Let's add some work for next tick so the worker cannot exit.
```

Comment:
nit

```suggestion
    // Add some work for next tick so the worker cannot exit.
```

The "let's" makes it sound like doing it for jollies.